### ISSUE TITLE
Code snippet freezes pyright

### ISSUE THREAD

**Issue by tomhampshire** [microsoft/pyright#4424]
Running pyright on the included code snippet causes pyright to run infinitely.

**Describe the bug**
Running pyright on the included code snippet causes pyright to run infinitely. The code itself has been minimised as much as possible, and is nonsensical.

**To Reproduce**
Put the code in a file and run `pyright` from the command line (uses pyright 1.1.287 on Arch Linux). Tested with Python 3.10.9 and Python 3.9.16.

**Expected behavior**
pyright should finish running.

**Screenshots or Code**

```python

def function_a():
    function_b([])


def function_b(input_list):
    output_tuple = ()

    for _, value in enumerate([]):
        if value is None:
            output_tuple += (None,)
            continue
        output_tuple += (input_list[value],)

    return output_tuple

```

**VS Code extension or command-line**
pyright 1.1.287 from the Arch Linux community/pyright 1.1.287-1

**Comment by erictraut** [comment]
Thanks for the bug report. I've found the underlying cause, but I'm still working out a fix. The issue is that every time through the loop, a new `tuple` type is produced for `output_tuple`. This gets unioned with the tuple types from the previous loop passes. The type never converges to a single type. It simply adds more and more subtypes to the union.

**Comment by erictraut** [comment]
I found a reasonable fix. This will be included in the next release.

**Comment by erictraut** [comment]
This is included in pyright 1.1.289, which I just published. It will also be included in a future release of pylance.

**Comment by tomhampshire** [comment]
Thanks for looking into and fixing this!
